File: src/20.py
# https://leetcode.com/problems/valid-palindrome/description/
class Solution:
    def isValid(self, s: str) -> bool:
        # the problem is to verify that all the paranthesis in the string are
        # balanced correctly
        # valid sets are '()', '{}', and '[]'
        #
        # these are the rules
        # 1. Open brackets must be closed by the same type of brackets.
        # 2. Open brackets must be closed in the correct order.
        # 3. Every close bracket has a corresponding open bracket of the same
        #    type.
        #
        # a stack is a pretty natural solution for this problem. as we add a new
        # character onto the stack, we can check if the last character added
        # is the corresponding character

        # use a list to simulate the stack
        # we can append() and pop() to add and remove to and from the stack
        stack = list()

        mapping = {
            ")": "(",
            "}": "{",
            "]": "[",
        }

        # loop over each of the characters
        for c in s:
            match c:
                # add the opening parenthesis
                case "(" | "{" | "[":
                    stack.append(c)
                # check the closing parenthesis
                case ")" | "}" | "]" if stack and stack[-1] == mapping[c]:
                    stack.pop()
                # otherwise, we must have missed the match
                case _:
                    return False

        # the stack should be empty!
        return not stack

File: src/cpp/20.cpp
#include <stack>
#include <string>
#include <unordered_map>

class Solution {
public:
  bool isValid(std::string s) {
    // in this problem, we are checking if a string of parenthesis is valid
    // - every opening ( has a closing )
    // - every opening { has a closing }
    // - every opening [ has a closing ]
    // moreover, open brackets are closed in the correct order

    std::stack<char> st;
    std::unordered_map<char, char> paren_mapping = {
        {')', '('},
        {'}', '{'},
        {']', '['},
    };

    for (const auto &c : s) {
      char prev;
      switch (c) {
      case '(':
      case '[':
      case '{':
        st.push(c); // push all opening parens
        break;
      case ')':
      case ']':
      case '}':
        char corres = paren_mapping[c];
        if (!st.empty() && st.top() == corres) {
          st.pop(); // pop the opening paren only if the current paren matches
        } else {
          return false;
        }
        break;
      }
    }

    // the stack must be empty at the end to be resolved
    return st.empty();
  }
};
